### PR TITLE
Add user defined literals for megabytes and kilobytes

### PR DESCRIPTION
**Pull request by JoeLoser**
I know that there is some developer overhead in knowing that these user-defined-literals exist in the codebase, but I felt the several usages of simple multiplications of `1024` combined with comments detailing the units justified introducing the two simple UDLs.

Note that I didn't introduce the whole slew of byte-like UDLs as I only saw widespread usage of kb and mb.

Requested reviewers:
@seelabs @HowardHinnant

### CODE REVIEW

**Comment by MarkusTeufelberger** [comment]
Aren't these technically https://en.wikipedia.org/wiki/Mebibyte s? Also all lowercase letters might be confused with bits.

File: src/ripple/basics/literals.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2011 The Bitcoin developers
```
[the remark this answers is not in the document]
Comment (JoeLoser) [reply src/ripple/basics/literals.h]: @seelabs it's a copy-paste error. I copied the header/copyright from `base_uint` not realizing that the ideas from that file *did* originate from bitcoin (https://github.com/ripple/rippled/blob/develop/src/ripple/basics/base_uint.h#L20). All I modified was the copyright year for Ripple Labs to be 2018.  This file is not used from what I know in bitcoin. I'll remove the accidental copyright wrt bitcoin -- sorry about that.

**Comment by JoeLoser** [comment]
@MarkusTeufelberger valid point as we use a power of 2 (1024) rather than 1000. These are technically mebibytes then.  As for the naming of all lowercase, I can see that concern as well.

I don't have strong opinions on the casing/spelling of these UDLs, so I'm happy to rename to something everyone agrees on. What are your thoughts, @HowardHinnant @seelabs?

**Comment by seelabs** [comment]
This PR allows us to replace code like: `maxBytes = 4*1024` with `maxBytes = 4_kb`. Here are the issues I'm thinking about:

1) The type return by `operator""` is `unsigned long long`. This changes the types of some expressions and we need to audit the type change (or cast to the original type).

2) I think of user defined literals as a way to easily a specific _type_, not a specific quantity. I.e. `"300"s` creates a `std::string` and `300s` creates a `chorno::duration`. `4_kb` creates an `unsigned long long`, which seems like a different usage.

3) Anytime we create a shallow abstraction like `operator""_kb` we have to weigh how often it'll be used and the clarity of the use. The concern is that a handful of uses means most developers won't be familiar with the abstraction, and they'll end up jumping to the definition of the abstraction anyway. I.e. everyone know what `maxBytes = 4*1024;` means. If we write that as `maxBytes = 4_kb` I end up looking up what `_kb` means and I'd rather it was written inline. I know you're aware of this as well.

I _do_ like the `maxBytes = 4_kb` form. I wanted to post this here to both solicit other opinions and to explain why I didn't just give this a quick thumbs up. I need to think though these issues myself first.

**Comment by seelabs** [comment]
@HowardHinnant On my system the following compiles without warning:
```
    {
        using namespace ripple;
        auto const i = -512_mb; (void) i;
        auto const j = -1*512_mb; (void) j;
        std::cout << "i: " << i << '\n';
        std::cout << "j: " << j << '\n';
        exit (0);
    }
```
And outputs:
```
i: 18446744073172680704
j: 18446744073172680704
```
I do think we need to audit the code with an eye to the type change if we take this change. Of course, a warning would be much better because it would prevent future misuses (I think we disable the signed/unsigned warnings, but I haven't dug into it).

**Comment by seelabs** [comment]
What do you think about using a template function instead: `template<class T> T megabytes(T){...}`? It resolves both the type change to `unsigned long long` and the issue that this is used to specify a quanity and not a new type.

So: replace `maxBytes = 4*1024;` with `maxBytes = kilobytes(4);`.

**Comment by nbougalis** [comment]
My _personal_ opinion is that this doesn’t really help elucidate the code. Perhaps it’s because I’m just not used to mentally parsing user-defined literals just yet, but between the novelty of UDLs (I know about the `<chrono>` stuff) and the awkward `_` prefix, I just see this as marginally detrimental. YMMV.

I do think we should look at this with a critical eye and decide whether, in the long term, it’s going to help the codebase.

Again, this is purely my personal opinion and I’m not speaking ex cathedra.

File: src/ripple/app/main/Application.cpp [hunk 2]
```diff
@@ -1143,11 +1142,11 @@ bool ApplicationImp::setup()
 
     getLedgerDB ().getSession ()
         << boost::str (boost::format ("PRAGMA cache_size=-%d;") %
-                        (config_->getSize (siLgrDBCache) * 1024));
+                        (config_->getSize (siLgrDBCache) * 1_kb));
```
Comment (seelabs) [inline src/ripple/app/main/Application.cpp]: I don't think I can give a thumbs up for this until we can come up with a resolution for the type change problem. Consider the code above. If someone changes that to move the minus sign out of the format and into the calculation, we're computing the wrong value. We'll certainlly catch those problems when reviewing this changeset when we all have the issue in mind. However, it's going to be an ongoing maintence issue to remember that `_kb` and `_mb` are `unsigned long long`.

**Comment by JoeLoser** [comment]
@seelabs I share your same concerns in your list of three items.  

> The return type of operator"" is unsigned long long. This changes the types of some expressions and we need to audit the type change (or cast to the original type).

Agreed. I thought we would have signed-conversion warnings on that would catch these subtle type issues and I would get flagged. As @HowardHinnant pointed out, neither one of us had build errors on OS X. But, we explicitly say `-Wno-sign-compare` in our builds (https://github.com/JoeLoser/rippled/blob/filesize_literals/Builds/CMake/CMakeFuncs.cmake#L657), so we would have to properly audit individual usages in this PR if we choose to keep pursuing it.

As a separate effort and PR, I think it would be good to fix all the culprit call sites such that we do need to ignore sign comparison warnings. As you know, they can lead to tricky, unexpected bugs.

> I think of user defined literals as a way to easily specify a type, not a specific quantity. I.e. "300"s creates a std::string and 300s creates a chorno::duration. 4_kb creates an unsigned long long, which seems like a different usage.

From my perspective, I think of all of those UDLs as creating values.  Now, whether those values form actual units (in the `Boost.Units` sense) and whether they are strong named types are what form the differences. In the chrono case, you have strong types and can form a taxonomy to do conversions on with duration. In the string case, not so much (string literal `_s` is just creating an unnamed value of type `std::string`) which I see as the similar thing in this case. Our return type is a weak type just like in the `_s` case.

**Comment by JoeLoser** [comment]
@seelabs I'd be fine with going down a function template approach to name these quantities; good recommendation! It's slightly more characters than the existing code, though it does hide some of the details of the calculations and makes intent explicit at the call sites.

@HowardHinnant @nbougalis are either of you opposed to going down a function template approach for this problem?

In a separate PR, when I get some time, I can dig into the sign comparison warnings and see how bad they are in cardinality.

**Comment by HowardHinnant** [comment]
I am not opposed to a function template approach.  I also don't see an urgent need for it.

**Comment by MarkusTeufelberger** [comment]
I'm still not convinced to use standardized SI prefixes for powers of 2 that are close to these numbers.

Ten kilobytes != ten kibibytes. instead of having these numbers explicitly stated, they now need to be looked up if they actually mean what they say (in the case of this PR they are simply the wrong unit).

It is a good rule of thumb to equate these numbers just like there are about 2 pounds per kg, 3 feet to a meter or 4 liters in a gallon. If they are used in code though I would expect `megabytes(2.75)` to be exactly 2.75 million bytes for example. This PR means these numbers would be about 5% off and I would guess that without a calculator it would take you a while to even know the exact number that this function would return. Even knowing immediately if that number is an integer value or if it would be rounded up/down takes a few more thought cycles.

In short: Please rename the functions to use a https://en.wikipedia.org/wiki/Binary_prefix

**Comment by nbougalis** [comment]
Merged as cd1c5a30ddacc140cde64c4ca4d4deaa7f345595

**Comment by nbougalis** [comment]
Sorry @MarkusTeufelberger: I didn't see your message until _after_ I had merged this as part of **1.2.0-b2**. Mea culpa. I'll make it a point of doing a final pass across all PRs to ensure there are no last-minute objections.

Also, in the future, if you could post a review that's marked as "Changes Requested" it would help, as it shows up with a red X instead of a pleasant green checkmark.

**Comment by MarkusTeufelberger** [comment]
Yeah, I wasn't sure if external reviews even help and it was more a comment that this complicates things more than they need to by using imprecise language. Overall I still think it is a net positive compared to "naked" numbers (maybe with a comment attached behind/above them).

**Comment by nbougalis** [comment]
The definitely help and we do appreciate them! Keep'em coming!
